### PR TITLE
Optimize `augment-vis` saved obj searching by adding arg to saved obj client

### LINKED ISSUE #4593: Increase `augment-vis` saved obj limits on a dashboard
By default, the saved object loader `findAll()` has a limit of 100 objects and doesn't implement pagination, as that is an open/incomplete item on OSD side. Because of this, there is a default limitation of 100 objects (associations) that we can show on a given dashboard when all of them are fetched.

Saved object loader `findAll()`: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/saved_objects/public/saved_object/saved_object_loader.ts#L122-L147
Augment vis saved object loader `findAll()` (just a `super` call): https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis.ts#L85-L97

A few options to improve this:
1. Add other search fields to `findAll()` call so we only fetch the objs for a particular vis ID. This is because right now, we are making this call for each visualization, fetching all of the same objs, and then doing post-filtering to only include ones that match the vis ID. This is inefficient and adds unnecessary network load
2. Increase the base size limit of 100 to something larger. Depending on what we want to support per visualization, and if the first option is completed first, perhaps 100 is sufficient.
3. Implement pagination at the base saved object loader level. 

Option 1 will likely be sufficient to solve the problems. If we end up wanting to support >100 associations per vis, we can look into Options 2/3

### PR DESCRIPTION
**Pull request by ohltyler**
### Description

This PR changes the fetching logic used in helper fns from fetching all objects and filtering by vis id -> fetching only objects that have the vis ID referenced via `has_reference` query param provided by the saved object `_find` api.

This byproduct of this change is much better performance and lower network load since only necessary objects are returned from the API call. Additionally, it removes the somewhat-hidden "limit" of 100 associations being visible/available on a dashboard. This is because before, when fetching associations for a visualization, it was only able to fetch a global 100 objects since all of the filtering was happening in post-processing instead of utilizing filtering provided by the API itself. Now, there is no limitation on a dashboard, since the limit (which is still 100 for now) is at a per-visualization level instead of a global per-dashboard level.

I had hit many problems when trying to pass a `has_references` object into the query, as the request would fail due to JSON parsing issues. I eventually found out that the existing code does not properly work for this one param, since it is the only one that is an `object`. By stringifying it, it's able to be read and parsed correctly. I think the culprit is that `SavedObjectFindOptions` is actually an interface coming from server-side (see [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/saved_objects/saved_objects_client.ts#L39)), such that an object rather than a stringified value makes more sense there. But in order to pass it _to_ server-side, it is passed via query params where it needs to be encoded correctly. And in fact, the only place where this field is even mentioned on public/client-side code is the test examples in the saved object client. I think the 2 options are 
1. Leave as-is and do the post-processing right before executing the query to keep client-side & server-side params the same, or 
2. Update the `hasReference` field for `SavedObjectFindOptions` as a string rather than an object, then I could remove the post-processing and force callers to format it into a JSON-stringified string.
I think I prefer 1 for simplification. Can add a comment about this as well.

- [x] Confirmed with AD/Alerting plugins that the deleted function `getAllAugmentVisSavedObjs` is not being used. Updated the one instance where it is used in core OSD.
- [x] Confirmed all saved obj client tests + all vis augmenter tests pass locally

### Issues Resolved

Closes #4593 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ohltyler** [comment]
Please see note about `has_reference` above, there's a few options available.

**Review by lezzago — DISMISSED** [review]

**Review by kavilla — CHANGES_REQUESTED** [review]
Seeing test failures.

**Comment by ohltyler** [comment]
> Seeing test failures.

addressed

**Comment by ohltyler** [comment]
I've confirmed on a dashboard with associations that this works as expected. For instance, if max limit is 3 objs fetched, and 2 visualizations have 3 associations (6 total), it still fetches the 3 visualizations for each, ensuring that filtering on vis ID works as expected.

File: src/plugins/vis_augmenter/public/actions/saved_object_delete_action.ts [hunk 1]
```diff
@@ -45,12 +45,16 @@ export class SavedObjectDeleteAction implements Action<SavedObjectDeleteContext>
...
+      const loader = getSavedAugmentVisLoader();
+      const augmentVisObjs = await getAugmentVisSavedObjs(savedObjectId, loader);
+      const augmentVisIdsToDelete = augmentVisObjs.map(
+        (augmentVisObj) => augmentVisObj.id as string
+      );
+
+      if (!isEmpty(augmentVisIdsToDelete)) loader.delete(augmentVisIdsToDelete);
+      // silently fail. this is because this is doing extra cleanup on objects unrelated
+      // to the user flow so we dont want to add confusing log lines or error toasts
+    } catch (e) {} // eslint-disable-line no-empty
```
Comment (ashwin-pc) [inline src/plugins/vis_augmenter/public/actions/saved_object_delete_action.ts:58]: You can at least log this error to the console.

**Review by ashwin-pc — COMMENTED** [review]
@ohltyler The change looks good to me. Just one callout, it looks like you are changing the saved client and how it works. This needs to be called out in the changelog for sure even if its not a breaking change

File: src/plugins/vis_augmenter/public/actions/saved_object_delete_action.ts (hunk 1 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/saved_object_delete_action.ts]: Added

**Review by ashwin-pc — APPROVED** [review]

**Review by lezzago — APPROVED** [review]

**Pull request merged**